### PR TITLE
Mark RuleWalker and ProgramAwareRuleWalker as deprecated

### LINKED ISSUE #4298: Mark RuleWalker, ProgramAwareRuleWalker, and ScopeAwareRuleWalker as deprecated
Recapping #2522: `RuleWalker`-based rules are slow, so it's generally preferable to use `applyWithFunction` instead of `applyWithWalker`.

if folks shouldn't use `RuleWalker`, it would be good to actively encourage new PRs not to. TSLint could mark them as `@deprecated` in their doc comments and put `// tslint:disable-next-line:deprecation` with links to issue(s) tracking converting them to `walk` functions.

Edit: ~~Two~~Three thumbs up from over a month - marking as accepting PRs 😊

### PR DESCRIPTION
**Pull request by Nazanin1369**
Fixes #4298

#### PR checklist

- [ x] Addresses an existing issue: #4298
- [ ] New feature, bugfix, or enhancement
- [ ] Includes tests
- [ ] Documentation update

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @Nazanin1369! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: src/language/walker/programAwareRuleWalker.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ import * as ts from "typescript";
 import { IOptions } from "../rule/rule";
 import { RuleWalker } from "./ruleWalker";
 
+// tslint:disable-next-line:deprecation
 export class ProgramAwareRuleWalker extends RuleWalker {
```
Comment (JoshuaKGoldberg) [inline src/language/walker/programAwareRuleWalker.ts]: Oh, I'd forgotten this exists! 😄

Could you please also mark this as deprecated @Nazanin1369? Its runtime works the same as `RuleWalker` _(but with the type checker and program)_, so it should be deprecated for the same reasons. I'll update the linked issue.

Comment (Nazanin1369) [reply src/language/walker/programAwareRuleWalker.ts]: Sure

Comment (Nazanin1369) [reply src/language/walker/programAwareRuleWalker.ts]: scopeAwareRuleWalker.ts already has the deprecated annotation.

**Comment by Nazanin1369** [comment]
@JoshuaKGoldberg  Looks like this is a breaking change with TS 3.0+. Do I need to change anything here?

**Comment by JoshuaKGoldberg** [comment]
Good catch - yes 😄

**Comment by adidahiya** [comment]
I don't consider this a breaking change, it's just a deprecation. Users might get errors with the `deprecation` lint rule but that's ok. What do you mean about TS 3.0+?

**Comment by JoshuaKGoldberg** [comment]
@adidahiya interesting, is it not? I would have thought build breaks from new `deprecation` rule violations to be considered breaking?

At any rate, I'd interpreted "TS 3.0+" to refer to the next major version of TSLint.

**Comment by adidahiya** [comment]
@JoshuaKGoldberg I see how it might be treated as breaking, but I've generally developed libraries without the constraint that deprecations can only occur at a major version change -- instead, they can occur at any minor version with the indication that the functionality will be removed in the next major version. At least, that's how we deprecate things in our internal libraries at Palantir. I find that TSLint sometimes introduces slightly stricter checks for rules across minor versions, so, in practice, users will sometimes see build breaks upon upgrading if they have lint failures reported as errors instead of warnings. You might argue that we shouldn't be afraid of major version bumps, we could just release 6.0 with this "break", move on to 7.0, etc, but given the current release cadence of TSLint, I'm fine with merging this change in the 5.x range.

**Pull request merged**